### PR TITLE
Fix sporadic hang in Mono.Debugger.Soft test suite on Windows.

### PR DESCRIPTION
**Pull request by lateralusX**
Mono.Debugger.Soft has low frequency hangs in InspectThreadSuspenedOnWaitOne test. This method launch the debuggee (dtest-app.exe), sets a breakpoint on a method just doing an infinite OS wait, validate that the breakpoint gets hit and resumes debuggee. After resume from breakpoint debuggee will hit the infinite OS wait and the debugger test will suspend and shutdown the process.

In order for mono debugger to handle the shutdown request it will suspend all managed threads using mono_thread_suspend_all_other_threads. This method uses a pooling schema until all threads have reported that they are suspended. Since one of the threads are doing an infinite OS wait we entered the land of APC (Asynchron Procedure Calls) used on Windows to alert waitable threads.

The reason for the hang is the fact that the OS won’t return from the wait until all queued APC’s have been executed. Since the loop sending async suspend requests to the threads mono_thread_suspend_all_other_threads can run faster (including suspending/resuming the thread that need to consume APC’s in the process) then the dequeue and execute of the queued APC’s, the wait won’t break, meaning that the thread won’t be able to set the mono suspend state causing the mono_thread_suspend_all_other_threads to run forever. This is highly timing dependent and will only reproduce sporadic, but I was able to isolate and do a solid repro locally while working on the fix.

The fix makes sure we don’t post a new APC for interrupt/abort requests if there is already one in flight for targeted thread. This will make sure we won’t flood the APC queue on Windows when requesting interrupts/aborts on waiting threads, giving them time to break out of OS wait and service interrupt/abort request.

### CODE REVIEW

**Comment by kumpera** [comment]
This is the wrong fix, we should change it on mono-threads and protect against overflow there.

An APC is similar to a signal and I believe we could use the same mechanics there as well.

**Comment by lateralusX** [comment]
Thanks for feedback! I put the fix at this level to reduce regression risk (I'm don't know all ins and outs of the thread implementation, especially not for none Windows platforms). I will investigate how to prevent the interrupt flooding in mono-threads.c or mono-threads-windows.c if it seems better to just limit the APC flooding during suspend requests.

**Comment by lateralusX** [comment]
@kumpera An alternative fix would be to not issue a new APC in mono-threads-windows.c, mono_threads_suspend_begin_async_suspend if the thread is already in an interrupt state as found out by mono_thread_info_is_interrupt_state. Doing the fix at this level will reduce the impact to fix the behavior of APC's on Windows causing the thread to hang and not affect any other platforms (that doesn't seem to suffer from the same problem). Would that be an acceptable fix to this problem?

**Comment by kumpera** [comment]
I though about something along the lines of having the APC callback clear a flag in MonoThreadInfo that guards whether we'd trigger an APC or not.

Do you think this could work?

**Comment by lateralusX** [comment]
@kumpera That's pretty much the behavior of the flag I reused in above suggestion, mono_thread_info_is_interrupt_state (info), it uses MonoThreadInfo's interrupt_token that is set when we issue a kernel interrupt (using set_interrupt_state) on a thread and it will be cleared when the targeting thread calls mono_thread_execute_interruption after interruption.

I could of course add yet another flag just handling the APC's in similar way using a different flag in MonoThreadInfo, if will be a little more precise in its nature and hopefully reduce potential side effects. We need to make sure that the APC flag is reset when we enter and leave any alertable OS wait. Would it be possible to clear that flag as part of MONO_ENTER_GC_SAFE and MONO_EXIT_GC_SAFE calls? Any call that could end up in OS wait calls should already be guarded with these macros anyway, correct?

So following that idea I could end up with something like this (windows only change due to the nature of APC's):

* Add new APC flag in MonoThreadInfo.
* Clear flag in MONO_ENTER_GC_SAFE and  MONO_EXIT_GC_SAFE.
* Only queue a new APC if the APC flag is cleared.
* Set flag when we queue an APC
* Move the APC call's into a shared method to make sure all current calls to QueueUserAPC handles the flag in a unified way.

Thoughts?

**Comment by lateralusX** [comment]
After discussion with @luhenry I ended up with a slightly modified version of above suggestions:

* Added new wait state flag in Windows version of MonoThreadInfo.
* Clear flag before entering a Windows alertable wait and after returning (atomic operations).
* Only queue a new APC if the wait state flag is not indicating a pending APC request.
* Set wait state flag when we queue an APC (atomic operation).
* Move the APC call's and aleratable waits into a shared methods to make sure all current calls to QueueUserAPC and WaitXXX handles the flag in a unified way.
* Added a new file mono-os-wait.h and a Windows implementation of it in mono-os-wait-win32.c in libmonoutils.
* Changed direct calls in codebase to Windows alertable waits like WaitForSingleObjectEx, WaitForMultipleObjectsEx etc, into calls to "wrappers" methods in mono-os-wait.h that will correctly handle the wait state flag.

**Comment by lateralusX** [comment]
Investigating regression in corlib on Windows.

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -4448,9 +4449,10 @@ mono_thread_execute_interruption (void)
 	}
 
 	/* this will consume pending APC calls */
-#ifdef HOST_WIN32
-	WaitForSingleObjectEx (GetCurrentThread(), 0, TRUE);
+#ifdef USE_WINDOWS_BACKEND
```
Comment (luhenry) [inline mono/metadata/threads.c]: Why this change?

File: mono/utils/mono-os-wait-win32.c [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+#include <mono/utils/mono-threads.h>
+#include <mono/utils/mono-threads-debug.h>
+
+#define THREAD_WAIT_INFO_CLEARED 0
+#define THREAD_WAIT_INFO_ALERTABLE_WAIT_SLOT 1
+#define THREAD_WAIT_INFO_PENDING_INTERRUPT_APC_SLOT 2
+#define THREAD_WAIT_INFO_PENDING_ABORT_APC_SLOT 4
+
+static inline void
+reqeust_interrupt (gpointer thread_info, HANDLE native_thread_handle, gint32 pending_apc_slot, PAPCFUNC apc_callback, DWORD tid)
```
Comment (luhenry) [inline mono/utils/mono-os-wait-win32.c]: there is a typo `reqeust`

File: mono/utils/mono-os-wait-win32.c [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	* NOTE, this code will executed regardless if thread is currently in an alertable wait or not.
+	* This is done to prevent races between interrupt/abort occurring just before thread enters an
+	* alertable wait. Threads entering waits already need to handle WAIT_IO_COMPLETION scenarios and
+	* if that happens due to a previous pending interrupt/abort APC, the calling thread should already
+	* have logic to restart the alertable wait operation.
+	*/
+	MonoThreadInfo *info = (MonoThreadInfo *)thread_info;
+	gboolean queue_apc = FALSE;
+
+	while (!queue_apc) {
```
Comment (luhenry) [inline mono/utils/mono-os-wait-win32.c]: The usual way in the runtime of formatting a `InterlockedCompareExchange` loop is the following:

```
gint old_wait_info, new_wait_info;
do {
    old_wait_info = info->thread_wait_info;
    if (old_wait_info & pending_apc_slot)
        return;

    new_wait_info = old_wait_info | pending_apc_slot;
} while (InterlockedCompareExchange (&info->thread_wait_info, new_wait_info, old_wait_info) != old_wait_info);

THREADS_INTERRUPT_DEBUG ("%06d - Interrupting/Aborting syscall in thread %06d", GetCurrentThreadId (), tid);
QueueUserAPC (apc_callback, native_thread_handle, (ULONG_PTR)NULL);
```

File: mono/utils/mono-os-wait-win32.c [hunk 4]
```diff
@@ -0,0 +1,210 @@
+#include <mono/utils/mono-os-wait.h>
+#include <mono/utils/mono-threads.h>
+#include <mono/utils/mono-threads-debug.h>
+
+#define THREAD_WAIT_INFO_CLEARED 0
```
Comment (luhenry) [inline mono/utils/mono-os-wait-win32.c]: It's usually defined in a enum, with `1 << 0`, `1 << 1`, etc.

**Comment by lateralusX** [comment]
@Kumpera, fine with these changes?

File: mono/utils/mono-os-wait-win32.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+*
+* Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#include <mono/utils/mono-os-wait.h>
+#include <mono/utils/mono-threads.h>
+#include <mono/utils/mono-threads-debug.h>
+
+enum ThreadWaitInfo {
+	THREAD_WAIT_INFO_CLEARED = 0,
```
Comment (kumpera) [inline mono/utils/mono-os-wait-win32.c]: I don't get why we need 3 APC queues, can't we get away with a single one?

Comment (lateralusX) [reply mono/utils/mono-os-wait-win32.c]: The first slot is used to indicate if a thread is currently in a OS wait call or not. Currently not used, but was put there for future use to detect if a thread is currently in an alertable OS wait call and potential optimize the APC queuing. The interrupt and abort slots was added since we had these two different APC types already in the code and used, so I didn't want to remove the possibility to differentiate them since that's what current code did, still making it possible to have both in flight at the same time.
